cpp/modmesh/serialization/SerializableItem.hpp

```diff
@@ -55,65 +56,265 @@ namespace detail
+
+/// State of the JSON parser.
+enum class JsonState
+{
+    Start,
+    ObjectKey,
+    Column,
+    Comma,
+    ObjectValue,
+    End,
+}; /* end enum class JsonState */
+
+/// Type of JSON token.
+enum class JsonType
+{
+    Object,
+    Array,
+    String,
+    Number,
+    Boolean,
+    Null,
+    Unknown,
+}; /* end enum class JsonType */
+
+struct JsonNode; // Forward declaration
+typedef std::unique_ptr<JsonNode> JsonNodePtr;
+typedef std::unordered_map<std::string, JsonNodePtr> JsonMap;
+typedef std::vector<JsonNodePtr> JsonArray;
+
+struct JsonNode
 {
-    if constexpr (std::is_base_of_v<SerializableItem, T>)
+    using JsonValue = std::variant<JsonMap, JsonArray, std::string>;
+
+    JsonType type;
+    JsonValue value;
+
+    JsonNode(JsonType type, const std::string & expression)
+        : type(type)
     {
-        return value.to_json(); /* recursive here */
+        parse(expression);
     }
-    else if constexpr (std::is_convertible_v<T, std::string>)
+
+private:
+
+    void parse(const std::string & expression)
     {
-        return "\"" + escape_string(value) + "\"";
+        if (type == JsonType::Object)
+        {
+            value = std::move(parse_object(expression));
+        }
+        else if (type == JsonType::Array)
+        {
+            value = std::move(parse_array(expression));
+        }
+        else
+        {
+            value = expression;
+        }
     }
-    else if constexpr (std::is_same_v<T, bool>)
+
+    static JsonMap parse_object(const std::string & json);
+
+    static JsonArray parse_array(const std::string & json);
+
+}; /* end struct JsonNode */
+
+/// Helper class for JSON serialization and deserialization.
+class JsonHelper
+{
+public:
+    template <typename T>
+    static std::string to_json_string(const T & value)
     {
-        return value ? "true" : "false";
+        if constexpr (std::is_base_of_v<SerializableItem, T>)
+        {
+            return value.to_json(); /* recursive here */
+        }
+        else if constexpr (std::is_convertible_v<T, std::string>)
+        {
+            return "\"" + escape_string(value) + "\"";
+        }
+        else if constexpr (std::is_same_v<T, bool>)
+        {
+            return value ? "true" : "false";
+        }
+        else if constexpr (is_specialization_of<std::vector, T>::value)
+        {
+            std::ostringstream oss;
+            oss << "[";
+            const char * separator = "";
+            for (const auto & item : value)
+            {
+                oss << separator << to_json_string(item); /* recursive here */
+                separator = ",";
+            }
+            oss << "]";
+            return oss.str();
+        }
+        else
+        {
+            return std::to_string(value);
+        }
     }
-    else if constexpr (is_specialization_of<std::vector, T>::value)
+
+    template <typename T>
+    static void from_json_string(const detail::JsonNodePtr & node, T & value)
```

Comment:
save the value into the class.

Reply:
I don't like the output argument.  But it looks like necessary evil here.  Is it so?  If yes, is it possible to make it private?

Reply:
And this function `from_json_string` is certainly too long for declaration.  Please make it outside the class declaration.

Reply:
> And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.

I put it in the class because L228 (when the parser doing recursion) uses the private method of `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

Reply:
> I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?

based on the current design, it is necessary.
What do you mean "private"? this is a helper class, and the function is static.
`SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

Reply:
In addition, since it's a helper class (or basically a helper function), I think the length should be fine? we have much longer helper class in `array_common.hpp`

Reply:
> > And this function from_json_string is certainly too long for declaration. Please make it outside the class declaration.
> 
> I put it in the class because L228 (when the parser doing recursion) uses the private method of `SerializableItem`, and that's why I declare `friend class detail::JsonHelper;`

The access is OK.  Long functions are not easy to read when putting in class declaration.  They should be taken outside the declaration while still made inline.

Reply:
> > I don't like the output argument. But it looks like necessary evil here. Is it so? If yes, is it possible to make it private?
> 
> based on the current design, it is necessary. What do you mean "private"? this is a helper class, and the function is static. `SerializableItem::from_json` calls this function, and feeds the class member throw the output argument.

It will be easier to maintain a function of output argument when it is private.  It does not matter the function to be static or not.  But if you use the class as a namespace for collecting static member functions and call them from other class, making them private is not realistic.

Please leave a code comment to warn maintainers of the output argument.  We should get rid of it in the future.

Reply:
@yungyuc  based on the current design:
in `from_json_string`:
```cpp
    else if constexpr (std::is_base_of_v<SerializableItem, T>)
    {
        if (node->type != detail::JsonType::Object)
        {
            throw std::runtime_error("Invalid JSON format: invalid object type.");
        }
        // NOLINTNEXTLINE(misc-no-recursion)
        value.from_json(std::get<JsonMap>(node->value)); /* recursive here */
    }
 ```
 
Note that we call `value`'s mutable method `from_json`. Therefore, we must pass the **mutable** reference of `value` as the argument of `from_json_string`.
 
 I don't think it can be an easy change to get rid of the output argument unless we change the whole design.
 
 Also,  I didn't fully understand why you wanted to remove the output argument. This function is live in the `detail` namespace, and the implementation is hidden under the macro, I don't think the user has chance to use this function.

Reply:
It's a good point that the static member function is in `detail`.  I missed it.  In this case I agree the output argument is not so severe.  Could you please add a simple code comment indicating the output arguments are bad for maintenance and should be fixed in the future?  (A concrete plan does not need to be mentioned.)

It's quite possible for a maintainer to move a useful helper class outside `detail`, in the future.  Then it will become an issue.